### PR TITLE
Fix downstream SIGBUS by adapting memory growth mechanism

### PR DESCRIPTION
**Pull request by ureeves**
The representation of memory is changed to make growth a cheap operation, as opposed to extremely expensive. Previously the memory was copied onto a temporary file with the new size, now growth happens onto an anonymous region of memory beyond the contents of the file, but still contiguous.

The diffing algorithm is slightly adapted to these changes, and made more performant.

### CODE REVIEW

File: piecrust/src/store/memory.rs [hunk 1]
```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
         VMMemory(Box::new(memory))
     }
 }
+
+/// The memory used by a contract.
+///
+/// It consists of one large mmap that doesn't reserve swap space - see
+/// [MAP_NORESERVE]. The memory region mapped will always be the maximum
+/// possible, since virtual address space is cheap, but the writable area is
+/// controlled bu `len`.
```
Comment (miloszm) [inline piecrust/src/store/memory.rs:190]: nit: `bu` should be `by`

File: piecrust/src/store/memory.rs [hunk 2]
```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
...
+    /// the file is mmapped copy-on-write, reserving a region as large as
+    /// `MAX_MEM_SIZE`. In the second step an overlapping region is mapped to
+    /// physical memory, at an offset of file length, using `MAP_FIXED`. This
+    /// ensures the memory is free to grow beyond the boundaries of the file
+    /// into physical memory, in one contiguous memory region.
+    ///
+    /// # Errors
+    /// If the file already exists, and it has a length that is not a multiple
+    /// of WASM_PAGE_SIZE, the function will error.
+    pub fn map<P: AsRef<Path>>(file: P) -> io::Result<Self> {
```
Comment (miloszm) [inline piecrust/src/store/memory.rs:263]: nit: this `pub` could go away

File: piecrust/src/store/memory.rs [hunk 3]
```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
...
+}
+
+impl MemoryMmap {
+    /// Creates a new memory, backed by the system's physical memory.
+    ///
+    /// Memories created like this always have `MIN_MEM_SIZE`, unless
+    /// [`grow_by`] is called.
+    ///
+    /// [`grow_by`]: MemoryMap::grow_by
+    pub fn new() -> io::Result<Self> {
```
Comment (miloszm) [inline piecrust/src/store/memory.rs:235]: nit: this `pub` could go away

**Review by miloszm — APPROVED** [review]
3 nits but otherwise LGTM

File: piecrust/src/store/diff.rs [hunk 4]
```diff
@@ -10,40 +10,72 @@ use std::io::{Read, Write};
...
-/// larger than it's current length.
+/// Applies the given patch on top of the new memory, using the old one as a
+/// base for the patch.
+///
+/// First, the growth delta is read from the patch, and if it's larger than zero
+/// both memories are grown. Then the patch is applied on top on the new memory,
+/// stemming from the contents of the old.
+///
+/// # Important
+/// The old memory should be discarded (dropped) after this operation.
```
Comment (herr-seppia) [inline piecrust/src/store/diff.rs:62]: If it's so important, did you consider to take the ownership of the old Memory?

File: piecrust/src/store/memory.rs [hunk 5]
```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
...
+        let new_len = self.len + delta_bytes;
+
+        if new_len > MAX_MEM_SIZE {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidInput,
+                format!("Tried growing past the maximum number of bytes: {MAX_MEM_SIZE}"),
+            ));
+        }
+
+        self.len = new_len;
```
Comment (herr-seppia) [inline piecrust/src/store/memory.rs:323]: Does it worth to check if the len is still a multiple of wasm page?

File: piecrust/src/store/memory.rs [hunk 6]
```diff
@@ -186,3 +181,288 @@ impl From<Memory> for VMMemory {
...
+    }
+
+    fn size(&self) -> Pages {
+        let pages = self.len / WASM_PAGE_SIZE - 1; // there is one guard page
+        Pages(pages as u32)
+    }
+
+    fn style(&self) -> MemoryStyle {
+        MemoryStyle::Static {
+            bound: Pages(WASM_MAX_PAGES),
```
Comment (herr-seppia) [inline piecrust/src/store/memory.rs:370]: Should be "-1" due to guard page?

**Review by herr-seppia — APPROVED** [review]
LGTM

File: piecrust/src/store/diff.rs [hunk 7]
```diff
@@ -10,40 +10,72 @@ use std::io::{Read, Write};
...
-    Bsdiff::new(old, new).compare(writer)
+    debug_assert!(
+        new_len >= old_len,
+        "Length of memories should strictly increase"
+    );
+
+    let delta_len = new_len - old_len;
+    let delta_len_bytes = delta_len.to_le_bytes();
+
+    writer.write_all(&delta_len_bytes)?;
```
Comment (fed-franz) [inline piecrust/src/store/diff.rs:43]: does it make sense to do this is delta_len is 0 ?

File: piecrust/src/store/memory.rs (hunk 6 above)
Comment (ureeves) [reply piecrust/src/store/memory.rs:370]: The WASM spec says that bound is *all mapped pages*. That's why I used the full memory size here.

File: piecrust/src/store/diff.rs (hunk 7 above)
Comment (ureeves) [reply piecrust/src/store/diff.rs:43]: We need to know in `patch` if we need to grow the memory or not, so either we put the actual length there, like before, or a diff length.

File: piecrust/src/store/memory.rs (hunk 5 above)
Comment (ureeves) [reply piecrust/src/store/memory.rs:323]: We could do this as a debug assertion, but I wouldn't waste runtime on release mode.

File: piecrust/src/store/diff.rs (hunk 4 above)
Comment (ureeves) [reply piecrust/src/store/diff.rs:62]: Omg, you're right :man_facepalming:

Comment (ureeves) [reply piecrust/src/store/diff.rs:62]: Well partially. The memory is just an `Arc` to an inner structure. This `Arc` should be deleted, and not used again.

Comment (herr-seppia) [reply piecrust/src/store/diff.rs:62]: Damn, I was almost there

**Pull request merged**